hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/catalog/HoodieHiveCatalog.java

```diff
@@ -397,17 +399,22 @@ public CatalogBaseTable getTable(ObjectPath tablePath) throws TableNotExistExcep
     String path = hiveTable.getSd().getLocation();
     Map<String, String> parameters = hiveTable.getParameters();
     Schema latestTableSchema = StreamerUtil.getLatestTableSchema(path, hiveConf);
+    String pkColumnsStr = parameters.get(FlinkOptions.RECORD_KEY_FIELD.key());
+    List<String> pkColumns = StringUtils.isNullOrEmpty(pkColumnsStr)
+        ? null : StringUtils.split(pkColumnsStr, ",");
     org.apache.flink.table.api.Schema schema;
     if (latestTableSchema != null) {
+      // if the table is initialized from spark, the write schema is nullable for pk columns.
+      DataType tableDataType = DataTypeUtils.ensureColumnsAsNonNullable(
```

Comment:
@danny0405 i actually think i've faced with issue in other place as well. Is there a reason why by default we mark all columns as nullable in Flink?

Reply:
It is a common behavior: a column is by default nullable if user does not declare the nullability in DDL. And for primary keys, they must be forced as non-nullable.

Flink would generate correct avro schema if the table was initialized from Flink app, what we fix here is a table created by Spark, so i guess, spark does not take the primary key constraint into nullability somewhere.

Reply:
Yeah, i don't think this in Spark DS we enforce invariant that primary-keys has to be non-null